sdcm/remote/remote_libssh_cmd_runner.py

```diff
@@ -0,0 +1,138 @@
+from .libssh2_client.exceptions import AuthenticationException, UnknownHostException, ConnectError, \
+    FailedToReadCommandOutput, CommandTimedOut, FailedToRunCommand, OpenChannelTimeout, SocketRecvError, \
+    UnexpectedExit, Failure
+from .libssh2_client.result import Result
+from .base import RetryableNetworkException
+from .remote_base import RemoteCmdRunnerBase
+
+
+class RemoteLibSSH2CmdRunner(RemoteCmdRunnerBase, ssh_transport='libssh2'):  # pylint: disable=too-many-instance-attributes
+    """Remoter that mimic RemoteCmdRunner, under the hood it runs libssh2 client, instead of paramiko
+    Main problem in libssh2 - is that it is not thread safe, we mitigate this problem by having
+      _connection_thread_map - a dictionary in which we bind thread to the libssh2 session.
+    Whenever remoter read self.connection, we return value from _connection_thread_map associated with current thread,
+      And if it is not there, we create it.
+    """
+    connection_thread_map = threading.local()
+    exception_unexpected = UnexpectedExit
+    exception_failure = Failure
+    exception_retryable = (
+        # Exceptions that are not signaling about
+        AuthenticationException, UnknownHostException, ConnectError, FailedToReadCommandOutput,
+        CommandTimedOut, FailedToRunCommand, OpenChannelTimeout, SocketRecvError, socket.timeout
+    )
+
+    @property
+    def connection(self) -> LibSSH2Client:
+        """
+        Map connection to current thread.
+        If there is no such thread, create it.
+        """
+        connection = getattr(self.connection_thread_map, str(id(self)), None)
+        if connection is None:
+            connection = self._create_connection()
+            setattr(self.connection_thread_map, str(id(self)), connection)
+        return connection
+
+    @connection.setter
+    def connection(self, connection: LibSSH2Client):
+        setattr(self.connection_thread_map, str(id(self)), connection)
+
+    def _create_connection(self) -> LibSSH2Client:
+        return LibSSH2Client(
+            host=self.hostname,
+            user=self.user,
+            port=self.port,
+            pkey=os.path.expanduser(self.key_file),
+            timings=Timings(keepalive_timeout=300, connect_timeout=self.connect_timeout)
+        )
+
+    @retrying(n=3, sleep_time=5, allowed_exceptions=(RetryableNetworkException,))
+    def run(self, cmd: str, timeout: Optional[float] = None, ignore_status: bool = False,  # pylint: disable=too-many-arguments
+            verbose: bool = True, new_session: bool = False, log_file: Optional[str] = None, retry: int = 1,
+            watchers: Optional[List[StreamWatcher]] = None) -> Result:
+
+        watchers = self._setup_watchers(verbose=verbose, log_file=log_file, additional_watchers=watchers)
+
+        @retrying(n=retry)
+        def _run():
+            try:
+                if verbose:
+                    self.log.debug('Running command "{}"...'.format(cmd))
```

Comment:
It's better to use lazy formatting for log messages.

Reply:
fixed

Reply:
fixed